# 範例 : 計程車費率預測
https://www.kaggle.com/c/new-york-city-taxi-fare-prediction
***
- 使用程車費率預測競賽練習時間欄位處理

In [1]:
# 做完特徵工程前的所有準備
import pandas as pd
import numpy as np
import datetime
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import GradientBoostingRegressor
import warnings
warnings.filterwarnings('ignore')

data_path = '../part02/'
df = pd.read_csv(data_path + 'taxi_data1.csv')

train_Y = df['fare_amount']
df = df.drop(['fare_amount'] , axis=1)
df.head()

pickup_datetime  pickup_longitude  pickup_latitude  \
0  2011-10-21 23:54:10 UTC        -73.990580        40.761071   
1  2015-02-03 10:42:03 UTC        -73.988403        40.723431   
2  2014-03-16 18:58:58 UTC        -74.015785        40.715110   
3  2009-06-13 16:10:54 UTC        -73.977322        40.787275   
4  2014-06-12 03:25:56 UTC        -73.989683        40.729717   

   dropoff_longitude  dropoff_latitude  passenger_count  
0         -73.981128         40.758634                2  
1         -73.989647         40.741695                1  
2         -74.012029         40.707888                2  
3         -73.958030         40.778838                3  
4         -73.982490         40.761887                3

In [2]:
# 時間特徵分解方式:使用datetime
#f=df['pickup_datetime'].apply(lambda x: datetime.datetime.strptime(x, '%W'))
df['pickup_datetime'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strptime(x, '%Y-%m-%d %H:%M:%S UTC'))
df['pickup_year'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%Y')).astype('int64')
df['pickup_month'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%m')).astype('int64')
df['pickup_day'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%d')).astype('int64')
df['pickup_hour'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%H')).astype('int64')
df['pickup_minute'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%M')).astype('int64')
df['pickup_second'] = df['pickup_datetime'].apply(lambda x: datetime.datetime.strftime(x, '%S')).astype('int64')
df.head()

pickup_datetime  pickup_longitude  pickup_latitude  dropoff_longitude  \
0 2011-10-21 23:54:10        -73.990580        40.761071         -73.981128   
1 2015-02-03 10:42:03        -73.988403        40.723431         -73.989647   
2 2014-03-16 18:58:58        -74.015785        40.715110         -74.012029   
3 2009-06-13 16:10:54        -73.977322        40.787275         -73.958030   
4 2014-06-12 03:25:56        -73.989683        40.729717         -73.982490   

   dropoff_latitude  passenger_count  pickup_year  pickup_month  pickup_day  \
0         40.758634                2         2011            10          21   
1         40.741695                1         2015             2           3   
2         40.707888                2         2014             3          16   
3         40.778838                3         2009             6          13   
4         40.761887                3         2014             6          12   

   pickup_hour  pickup_minute  pickup_second  
0           23             54             10  
1           10             42              3  
2           18             58             58  
3           16             10             54  
4            3             25             56

In [3]:
# 將結果使用線性迴歸 / 梯度提升樹分別看結果
df_temp = df.drop(['pickup_datetime'] , axis=1)
scaler = MinMaxScaler()
train_X = scaler.fit_transform(df_temp)
Linear = LinearRegression()
print(f'Linear Reg Score : {cross_val_score(Linear, train_X, train_Y, cv=5).mean()}')
GDBT = GradientBoostingRegressor()
print(f'Gradient Boosting Reg Score : {cross_val_score(GDBT, train_X, train_Y, cv=5).mean()}')

Linear Reg Score : 0.026876871475636888
Gradient Boosting Reg Score : 0.7121425474580371


# 作業1
* 對照範例，試著加入星期幾 (day of week) 與第幾周 (week of year) 這兩項特徵，  
看看結果會比原本只有時間特徵分解的結果更好或更差?

In [4]:
# 加入星期幾與第幾周兩個特徵
"""
Your Code Here
"""

#help(datetime)
#df['pickup_weekdays']=df['pickup_datetime'].apply(lambda x : datetime.date.isoweekday(x))
df['pickup_weeks']=df['pickup_datetime'].apply(lambda x: datetime.date.isocalendar(x)[1])


In [5]:
# 將結果使用線性迴歸 / 梯度提升樹分別看結果
df_temp = df.drop(['pickup_datetime'] , axis=1)
train_X = scaler.fit_transform(df_temp)
print(f'Linear Reg Score : {cross_val_score(Linear, train_X, train_Y, cv=5).mean()}')
print(f'Gradient Boosting Reg Score : {cross_val_score(GDBT, train_X, train_Y, cv=5).mean()}')


Linear Reg Score : 0.02678995634489607
Gradient Boosting Reg Score : 0.7168874904825966


In [6]:
# 加上"日週期"特徵 (參考講義"週期循環特徵")
import math
df['day_cycle'] = df['pickup_hour']/12 + df['pickup_minute']/720 + df['pickup_second']/43200
df['day_cycle'] = df['day_cycle'].map(lambda x:math.sin(x*math.pi))
df.head()

pickup_datetime  pickup_longitude  pickup_latitude  dropoff_longitude  \
0 2011-10-21 23:54:10        -73.990580        40.761071         -73.981128   
1 2015-02-03 10:42:03        -73.988403        40.723431         -73.989647   
2 2014-03-16 18:58:58        -74.015785        40.715110         -74.012029   
3 2009-06-13 16:10:54        -73.977322        40.787275         -73.958030   
4 2014-06-12 03:25:56        -73.989683        40.729717         -73.982490   

   dropoff_latitude  passenger_count  pickup_year  pickup_month  pickup_day  \
0         40.758634                2         2011            10          21   
1         40.741695                1         2015             2           3   
2         40.707888                2         2014             3          16   
3         40.778838                3         2009             6          13   
4         40.761887                3         2014             6          12   

   pickup_hour  pickup_minute  pickup_second  pickup_weeks  day_cycle  
0           23             54             10            42  -0.025450  
1           10             42              3             6   0.333601  
2           18             58             58            11  -0.967083  
3           16             10             54            24  -0.888817  
4            3             25             56            24   0.782427

In [7]:
# 將結果使用線性迴歸 / 梯度提升樹分別看結果
df_temp = df.drop(['pickup_datetime'] , axis=1)
train_X = scaler.fit_transform(df_temp)
print(f'Linear Reg Score : {cross_val_score(Linear, train_X, train_Y, cv=5).mean()}')
print(f'Gradient Boosting Reg Score : {cross_val_score(GDBT, train_X, train_Y, cv=5).mean()}')

Linear Reg Score : 0.026326928573012175
Gradient Boosting Reg Score : 0.7143111514625058


# 作業2
* 對照範例的日週期效果，試著參考投影片完成年週期與周週期的特徵 (也可以用你自己想到的方式)，  
看看結果會比範例中的結果更好或更差?

    前述的週期所需數值都可由時間欄欄位組成, 
    但還⾸首尾相接 因此週期特徵還需以正弦函數( sin )或餘弦函數( cos )加以組合 
    例例如 : 
    週期 ( 正 : 冷 / 負 : 熱 )                        
    cos((月/6 + 日/180 )π)           

    周週期 ( 正 : 精神飽滿 / 負 : 疲倦 )                        
    sin((星期幾/3.5 + 小時/84 )π)           

    日週期 ( 正 : 精神飽滿 / 負 : 疲倦 )                       
    sin((小時/12 + 分/720 + 秒/43200 )π) 

In [8]:
# 加上"年週期"與"周週期"特徵
"""
Your Code Here
"""
df.head()
df['pickup_weekdays']=df['pickup_datetime'].apply(lambda x : datetime.date.isoweekday(x))
df['year_cycle']=df['pickup_month']/6 +df['pickup_day']/180
df['year_cycle']=df['year_cycle'].map(lambda x: math.cos(x*math.pi))

df['week_cycle']=df['pickup_weekdays']/3.5 + df['pickup_hour']/84
df['week_cycle']=df['week_cycle'].map(lambda x: math.sin(x*math.pi))


In [9]:
# 將結果使用線性迴歸 / 梯度提升樹分別看結果
df_temp = df.drop(['pickup_datetime'] , axis=1)
train_X = scaler.fit_transform(df_temp)
print(f'Linear Reg Score : {cross_val_score(Linear, train_X, train_Y, cv=5).mean()}')
print(f'Gradient Boosting Reg Score : {cross_val_score(GDBT, train_X, train_Y, cv=5).mean()}')

Linear Reg Score : 0.026001835563946085
Gradient Boosting Reg Score : 0.709496498992121


# Results

    # w/o hw_1 result:0.026876871475636888, 0.7104308578324505
    # add df['pickup_weekdays']= 0.02617125367979578, 0.7079522145956696  getting worse
    # add df['pickup_weeks']= 0.02678995634489607, 0.7172880816243007
    # add df['pickup_weekdays'], and df['pickup_weeks'] 0.026080352293792554 0.7162446509653912
    # add year, and weekly cycle information, getting worse... 